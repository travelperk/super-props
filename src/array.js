import { error } from "./utils";
import { randomNumber } from "./utils";

export default function array(
  shape,
  {
    nullable = false,
    allowEmpty = true,
    minLength = 0,
    maxLength = Infinity,
    unique = false
  } = {}
) {
  function type(props, propName, componentName) {
    const value = props[propName];
    if (value == null && nullable) return;

    if (!Array.isArray(value))
      throw error(
        `Property \`${propName}\` of \`${componentName}\` is not an array.`
      );

    if (value.length === 0 && !allowEmpty)
      throw error(
        `Property \`${propName}\` of \`${componentName}\` can't be an empty array.`
      );

    if (value.length < minLength)
      throw error(
        `Property \`${propName}\` of \`${componentName}\` can't be shorter than ${minLength}.`
      );

    if (value.length > maxLength)
      throw error(
        `Property \`${propName}\` of \`${componentName}\` can't be longer than ${maxLength}.`
      );

    value.forEach((item, index) => {
      shape({ [index]: item }, index, `${componentName}[]`);
    });

    if (unique && value.length !== Array.from(new Set(value)).length)
      throw error(
        `Property \`${propName}\` of \`${componentName}\` has repeated values.`
      );
  }

  type.toString = function() {
    return `array(${shape})`;
  };

  type.make = function() {
    const length = randomNumber({
      min: minLength,
      max: maxLength === Infinity ? 10 : maxLength,
      integer: true
    });

    const generatedArray = [];
    for (let i = 0; i < length; i++) {
      let element;
      do {
        element = shape.make();
      } while (unique && isElementInArray(element, generatedArray));
      generatedArray.push(element);
    }
    return generatedArray;
  };

  return type;
}

function isElementInArray(element, array) {
  return array.some(e => e === element);
}
