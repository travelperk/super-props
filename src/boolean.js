import { error, choose } from "./utils";

export default function boolean({ nullable = false } = {}) {
  function type(props, propName, componentName) {
    const value = props[propName];
    if (value == null && nullable) return;

    if (typeof value !== "boolean")
      throw error(
        `Property \`${propName}\` of \`${componentName}\` is not a boolean.`
      );
  }

  type.toString = function() {
    return "boolean()";
  };

  type.make = function make() {
    return choose([true, false]);
  };

  return type;
}
