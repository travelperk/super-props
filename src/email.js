import { string } from ".";

export default function email(options) {
  const pattern = /^[a-z0-9!#$%&'*+/=?^_`{|}~.-]+@[a-z0-9-]+(\.[a-z0-9-]+)*$/i;

  const type = string({ ...options, pattern });

  type.toString = function() {
    return "string()";
  };

  return type;
}
