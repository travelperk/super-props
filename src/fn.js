import { error } from "./utils";

export default function fn({ nullable = false } = {}) {
  function type(props, propName, componentName) {
    const value = props[propName];
    if (value == null && nullable) return;

    if (typeof value !== "function")
      throw error(
        `Property \`${propName}\` of \`${componentName}\` is not a function.`
      );
  }

  type.toString = function() {
    return "fn()";
  };

  return type;
}
