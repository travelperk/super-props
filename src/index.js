export { default as boolean } from "./boolean";
export { default as number } from "./number";
