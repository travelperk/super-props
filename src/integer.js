import number from "./number";

export default function integer(opts = {}) {
  const type = number({ integer: true, ...opts });

  type.toString = function() {
    return "integer()";
  };

  return type;
}
