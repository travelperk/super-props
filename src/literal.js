import { boolean } from ".";
import { error } from "./utils";

export default function value(literalValue) {
  if (typeof literalValue === "object" && literalValue !== null)
    throw new Error("literal value can't be an object");

  if (typeof literalValue === "function")
    throw new Error("literal value can't be a function");

  function type(props, propName, componentName) {
    const value = props[propName];

    if (value !== literalValue)
      throw error(
        `Property \`${propName}.\` of \`${componentName}\` was supposed to be \`${literalValue}\` but was \`${value}\`.`
      );
  }

  type.make = function() {
    return literalValue;
  };

  type.toString = function() {
    return `literal(${literalValue})`;
  };

  return type;
}
