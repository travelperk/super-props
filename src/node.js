import PropTypes from "prop-types";
import { error } from "./utils";

export default function node({ nullable = false } = {}) {
  function type(props, propName, componentName) {
    const value = props[propName];
    if (value === null) return;

    const validator = nullable ? PropTypes.node : PropTypes.node.isRequired;

    const validationError = validator(
      props,
      propName,
      componentName,
      "",
      "",
      "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED"
    );

    if (validationError) throw new error(validationError.message);
  }

  type.toString = function() {
    return "node()";
  };

  return type;
}
