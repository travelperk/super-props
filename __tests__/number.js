import { number } from "../src";

describe("number()", () => {
  describe(".make()", () => {
    it("should generate a random number value", () => {
      expect(typeof number().make()).toBe("number");
      expect(number({ min: 0 }).make()).toBeGreaterThanOrEqual(0);
      const integer = number({ integer: true }).make();
      expect(integer).toBe(Math.round(integer));
    });
  });

  it.each`
    value        | options                | shouldThrow
    ${1}         | ${{}}                  | ${false}
    ${1}         | ${{ integer: true }}   | ${false}
    ${1}         | ${{ integer: false }}  | ${false}
    ${1.1}       | ${{ integer: false }}  | ${false}
    ${1.1}       | ${{ integer: true }}   | ${true}
    ${"1"}       | ${{}}                  | ${true}
    ${0}         | ${{ min: 1 }}          | ${true}
    ${0}         | ${{ min: 0 }}          | ${false}
    ${0}         | ${{ max: 0 }}          | ${false}
    ${0}         | ${{ max: -1 }}         | ${true}
    ${0}         | ${{ min: -1, max: 1 }} | ${false}
    ${0}         | ${{ min: 1, max: 2 }}  | ${true}
    ${undefined} | ${{ nullable: true }}  | ${false}
    ${undefined} | ${{ nullable: false }} | ${true}
    ${null}      | ${{ nullable: true }}  | ${false}
    ${null}      | ${{ nullable: false }} | ${true}
  `(
    "should generate a validator for number values with options $options and value $value",
    ({ value, options, shouldThrow }) => {
      const expectation = expect(() =>
        number(options)({ value }, "value", "Component")
      );
      shouldThrow
        ? expectation.toThrowErrorMatchingSnapshot()
        : expectation.not.toThrow();
    }
  );
});
