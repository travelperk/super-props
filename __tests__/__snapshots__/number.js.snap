// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`number() should generate a validator for number values with options {"integer": true} and value 1.1 1`] = `"[TKType] Property \`value\` of \`Component\` is not an integer."`;

exports[`number() should generate a validator for number values with options {"max": -1} and value 0 1`] = `"[TKType] Property \`value\` of \`Component\` is supposed to be smaller than -1. It was 0."`;

exports[`number() should generate a validator for number values with options {"max": 2, "min": 1} and value 0 1`] = `"[TKType] Property \`value\` of \`Component\` is supposed to be bigger than 1. It was 0."`;

exports[`number() should generate a validator for number values with options {"min": 1} and value 0 1`] = `"[TKType] Property \`value\` of \`Component\` is supposed to be bigger than 1. It was 0."`;

exports[`number() should generate a validator for number values with options {"nullable": false} and value null 1`] = `"[TKType] Property \`value\` of \`Component\` is not a number."`;

exports[`number() should generate a validator for number values with options {"nullable": false} and value undefined 1`] = `"[TKType] Property \`value\` of \`Component\` is not a number."`;

exports[`number() should generate a validator for number values with options {} and value 1 1`] = `"[TKType] Property \`value\` of \`Component\` is not a number."`;
